pkg/broker/ensure.go

```diff
@@ -64,5 +66,35 @@ func Ensure(config *rest.Config, ipsecPSKBytes int) error {
 	if err != nil && !apierrors.IsAlreadyExists(err) {
 		return fmt.Errorf("error creating the IPSEC PSK secret: %s", err)
 	}
-	return nil
+
+	return waitForClientToken(clientset)
+
+}
+
+func waitForClientToken(clientset *kubernetes.Clientset) error {
+
+	// wait for the client token to be ready, while implementing
+	// exponential backoff pattern, it will wait a total of:
+	// sum(n=0..9, 1.2^n * 5) seconds, = 130 seconds
+
+	backoff := wait.Backoff{
+		Steps:    10,
+		Duration: 5 * time.Second,
+		Factor:   1.2,
+		Jitter:   1,
+	}
+
+	var lastErr error
+	err := wait.ExponentialBackoff(backoff, func() (bool, error) {
+		_, lastErr = GetClientTokenSecret(clientset, SubmarinerBrokerNamespace)
+		if lastErr != nil {
+			return false, nil
```

Comment:
Don’t we want

    return false, lastErr

here? Or would that involve decoding the error to figure out if we want to retry?

Reply:
if we do that, the wait loop stops and exits https://github.com/kubernetes/apimachinery/blob/master/pkg/util/wait/wait.go#L292

So if we have an error because the token does not exist yet, we need to continue.

Reply:
I  must admit I also found it weird.

Reply:
Right, that’s what I meant by decoding the error: some errors indicate that we should continue (“token not found” etc.), others indicate that there’s no point waiting further (if the clientset failed to communicate with the target). Let’s leave it like that for now, at least we won’t miss errors!